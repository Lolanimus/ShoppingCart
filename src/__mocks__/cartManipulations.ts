import { vi } from "vitest";

export default {
    rewriteCart: vi.fn()
};