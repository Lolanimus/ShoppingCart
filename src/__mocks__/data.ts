const urls = ["https://fakestoreapi.com/products/1", "https://fakestoreapi.com/products/5", "https://fakestoreapi.com/products/15"];

const contents = [
    {
        id: 1,
        title: "Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops",
        price: 109.95,
        description: "Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday",
        category: "men's clothing",
        image: "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg",
        rating: {
            rate: 3.9,
            count: 120
        }
    },
    {
        id: 5,
        title: "John Hardy Women's Legends Naga Gold & Silver Dragon Station Chain Bracelet",
        price: 695,
        description: "From our Legends Collection, the Naga was inspired by the mythical water dragon that protects the ocean's pearl. Wear facing inward to be bestowed with love and abundance, or outward for protection.",
        category: 'jewelery',
        image: 'https://fakestoreapi.com/img/71pWzhdJNwL._AC_UL640_QL65_ML3_.jpg',
        rating: { 
            rate: 4.6, count: 400 
        }
    },
    {
        id: 15,
        title: "BIYLACLESEN Women's 3-in-1 Snowboard Jacket Winter Coats",
        price: 56.99,
        description: 'Note:The Jackets is US standard size, Please choose size as your usual wear Material: 100% Polyester; Detachable Liner Fabric: Warm Fleece. Detachable Functional Liner: Skin Friendly, Lightweigt and Warm.Stand Collar Liner jacket, keep you warm in cold weather. Zippered Pockets: 2 Zippered Hand Pockets, 2 Zippered Pockets on Chest (enough to keep cards or keys)and 1 Hidden Pocket Inside.Zippered Hand Pockets and Hidden Pocket keep your things secure. Humanized Design: Adjustable and Detachable Hood and Adjustable cuff to prevent the wind and water,for a comfortable fit. 3 in 1 Detachable Design provide more convenience, you can separate the coat and inner as needed, or wear it together. It is suitable for different season and help you adapt to different climates',
        category: "women's clothing",
        image: 'https://fakestoreapi.com/img/51Y5NI-I5jL._AC_UX679_.jpg',
        rating: { 
            rate: 2.6, 
            count: 235 
        }
    }
];

export {
    urls,
    contents
}