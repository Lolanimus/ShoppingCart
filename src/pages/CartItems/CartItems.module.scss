.cartItem {
    display: grid;
    grid-template-columns: 1fr 4fr;
    border-bottom: 1px solid var(--light-grey);
    padding: 24px 0;

    > section {
        align-self: center;
    }

    :has(> ol) {
        display: grid;
    }

    form {
        height: 100%;
        display: grid;
        grid-template-rows: repeat(3, 1fr);
    }

    .title {
        font-size: 1.75rem;
        font-weight: bold;
        padding-top: 12px;
    }

    .size {
        font-size: 1.5rem;
        align-self: center;
        align-items: center;
        display: grid;
        grid-template-columns: 1fr 1fr;
        
        :first-child {
            > span {
                font-size: 1.5rem;
                font-weight: 500;    
            }
            
            > div {
                font-size: 1.5rem;
                font-weight: 500;
            }
        }

        :nth-child(2) {
            justify-self: end;
        }
    }

    .itemCartSettings {
        display: grid;
        grid-template-columns: 1fr 1fr;
        align-content: flex-end;
    }

    .deleteBtn {
        border: none;
        background-color: white;
        padding-left: 0;
        padding-bottom: 0;
    }

    .price {
        font-weight: bolder;
        font-size: 24px;
        justify-self: end;
        padding: 12px;
    }
}

@media (max-width: 800px) {
    .cartItem {
        grid-template-columns: auto;
        border-bottom: 1px solid var(--light-grey) !important;
        padding: 0;
        margin-bottom: 24px;
        > ol {
            padding: 0;
        }

        > ol > form {
            grid-template-columns: 0.5fr 0.5fr;
            grid-template-rows: 0.5fr 0.5fr;
            gap: 12px;

            .size {
                grid-template-columns: auto;
                gap: 8px;
                
                :nth-child(2) {
                    justify-self: unset;
                }

                > div {
                    display: flex;

                    > button:first-child {
                        margin-left: 0;
                    }
                }

                > div:first-child > span {
                    margin-left: 4px;
                }
            }

            .price {
                padding: 0;
            }

            .deleteBtn {
                justify-self: flex-end;
            }
        }
    }

    :last-child {
        border: none !important;
    }
}