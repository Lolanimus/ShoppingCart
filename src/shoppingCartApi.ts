import { setCart, getCart } from "./cart";

const fetchData = async (url: string): Promise<CatalogArr> => {
    let res: CatalogArr = [];
    
    try {
        const data = await fetch(url)!;
        if(data.ok && data.status == 200) {
            const json: CatalogArr | CatalogObj = await data.json()!;
            res = json !instanceof Array ? json : [json];
        } else {
            throw { code: data.status, message: data.statusText };
        }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    } catch (error: any) {
        console.error("Error: " + error);
    }

    return res;
}

const getCatalog = (sex: string, catalog: CatalogArr) => {
    const returnedCatalog: CatalogArr = [];
    const category = sex + "'s clothing";
    catalog.forEach((obj) => {
        (obj.category === category || obj.category === "jewelery") ? returnedCatalog.push(obj) : null;
    });
    return returnedCatalog;
};

const getTotalPrice = () => {
    let totalPrice = 0;
    const cart = getCart();
    cart.forEach((item) => {
        totalPrice += item.quantity * item.price;
    })
    return totalPrice;
}

const addToCart = (item: CatalogObj, size?: string) => {
    const cartItem = {
        ...item,
        quantity: 1,
        size,
    };
    const tempCart = getCart();
    tempCart.push(cartItem);
    setCart(tempCart);
}

const deleteFromCart = (productId: number) => {
    const tempCart = getCart();
    tempCart.forEach((product, i) => {
        product.id === productId && tempCart.splice(i, 1);
    })
    setCart(tempCart);
}

const incrementQuantityCart = (productId: number, isIncrement: boolean) => {
    const tempCart = getCart();
    tempCart.forEach(product => {
        product.id === productId && (isIncrement ? product.quantity += 1 : (product.quantity > 1 ? product.quantity -= 1 : deleteFromCart(productId)))
    })
    setCart(tempCart);
}

const clearCart = () => {
    setCart([]);
}

export { getCatalog, getTotalPrice, addToCart, deleteFromCart, incrementQuantityCart, clearCart, fetchData };
