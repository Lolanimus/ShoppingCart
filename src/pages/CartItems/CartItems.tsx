import { CartItemProps } from "./CartItems.test";

const CartItems: React.FC<CartItemProps> = ({cartItems}) => {
    console.log(cartItems);
    const result = cartItems.length > 0 ? (
        <ul>
            {cartItems.map(item => (
                <li key={item.id}>
                    <section id="itemImg">
                        <img src={item.image} alt={item.title} />
                    </section>
                    <aside id="itemSettings">
                        <ul>
                            <li>
                                <span>{item.title}</span>
                            </li>
                            <li>
                                <span>{item.size}</span>
                            </li>
                            <li>
                                {/* <QuantityChanger /> */}
                            </li>
                            <li>
                                <span>{item.price}</span>
                            </li>
                            <button>delete</button>
                        </ul>
                    </aside>
                </li>
            ))}
        </ul>
    ) : (
        <span>There are no items in your cart yet...</span>
    )

    return result;
}

export default CartItems;