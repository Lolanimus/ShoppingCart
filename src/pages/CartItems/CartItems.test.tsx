import { render, within } from '@testing-library/react';
import { expect, describe, it } from 'vitest';
import CartItems from './CartItems';
import { addToCart } from '../../shoppingCartApi';
import * as data from "../../__mocks__/data"

describe("CartItems", () => {
  it("renders correctly(with size specified)", () => {  
    const itemIndex = 2;
    data.contents.forEach((obj) => {
      addToCart(obj, "m");
    })
    render(<CartItems />);
    const items = document.querySelectorAll(`ul > li`)!;
    const item = items[itemIndex] as HTMLElement;
    const img = within(item).getByRole("img");
    expect(img.getAttribute('src')).toBe(data.contents[itemIndex].image);
    const itemInfo = within(item).getByRole("list");
    const itemTitle = within(itemInfo).getByTestId("title");
    expect(itemTitle.textContent).toBe(data.contents[itemIndex].title);
    const itemSize = within(itemInfo).getByTestId("size");
    expect(itemSize.textContent).toBe("m");
    const itemQuantityChanger = within(itemInfo).getByTestId("quantityDiv");
    expect(itemQuantityChanger).toBeInTheDocument();
    const itemPrice = within(itemInfo).getByTestId("price");
    expect(itemPrice.textContent).toBe("" + data.contents[itemIndex].price);
    const itemDeleteButton = within(itemInfo).getByRole("button", {name: "Delete"});
    expect(itemDeleteButton).toBeInTheDocument();
  })
})