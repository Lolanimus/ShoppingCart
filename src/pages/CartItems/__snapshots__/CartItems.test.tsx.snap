// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CartItems > renders correctly 1`] = `
<DocumentFragment>
  <span>
    There are no items in your cart yet...
  </span>
</DocumentFragment>
`;
